### PR TITLE
Add support for foreign libraries

### PR DESCRIPTION
**Pull request by snowleopard**
This is an implementation of the proposal #2650. 

EDIT: Ready for review now.

### CODE REVIEW

File: src/dune/dune_env.ml [hunk 1]
```diff
@@ -8,11 +8,12 @@ module Stanza = struct
   let c_flags ~since =
     let check =
       Option.map since ~f:(fun since ->
-          Dune_lang.Syntax.since Stanza.syntax since)
+          Dune_lang.Syntax.deprecated_in Stanza.syntax (2, 0)
```
Comment (diml) [inline src/dune/dune_env.ml]: Let's change that to `deleted_in`

File: src/dune/dune_file.ml [hunk 2]
```diff
@@ -577,18 +472,51 @@ module Buildable = struct
     }
 
   let decode ~since_c ~allow_re_export =
+    let use_foreign =
+      Dune_lang.Syntax.deprecated_in Stanza.syntax (2, 0)
+        ~extra_info:"Use the (foreign_stubs ...) stanza instead."
```
Comment (diml) [inline src/dune/dune_file.ml]: s/stanza/field/

"stanza" is for toplevel things

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/dune/dune_env.ml (hunk 1 above)
Comment (snowleopard) [reply src/dune/dune_env.ml]: Done

File: src/dune/dune_file.ml (hunk 2 above)
Comment (snowleopard) [reply src/dune/dune_file.ml]: Done

**Comment by aalekseyev** [comment]
I added a couple of CRs which I think we should resolve before Dune 2.0, but we can do them in separate PRs.

Specifically, it's:

- make source foreign file conflict detection no longer complain about unused source files
- static/dynamic linking configurability in foreign_library stanza

**Review by aalekseyev — APPROVED** [review]

**Pull request merged**